### PR TITLE
Implements setup for research tool phasar

### PR DESCRIPTION
**Pull request by vulder**

### CODE REVIEW

File: varats/tools/research_tools/phasar.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                self, "phasar",
+                "https://github.com/secure-software-engineering/phasar.git",
+                "origin", "phasar"
+            )
+        ]
+        super().__init__(base_dir, sub_projects)
+
+    def checkout_phasar_version(self, use_dev_branche: bool) -> None:
+        """
+        Checkout out a specific version of VaRA.
```
Comment (boehmseb) [inline varats/tools/research_tools/phasar.py]:
```suggestion
        Checkout out a specific version of phasar.
```

File: varats/tools/research_tools/phasar.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        sub_projects = [
+            SubProject(
+                self, "phasar",
+                "https://github.com/secure-software-engineering/phasar.git",
+                "origin", "phasar"
+            )
+        ]
+        super().__init__(base_dir, sub_projects)
+
+    def checkout_phasar_version(self, use_dev_branche: bool) -> None:
```
Comment (boehmseb) [inline varats/tools/research_tools/phasar.py]: `use_dev_branche` -> `use_dev_branch`

File: varats/tools/research_tools/phasar.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    @staticmethod
+    def has_install_location() -> bool:
+        """Checks if a install location of the research tool is configured."""
+        return vara_cfg()["phasar"]["install_dir"].value is not None
+
+    @check_required_args(["install_prefix", "version"])
+    def setup(self, source_folder: tp.Optional[Path], **kwargs: tp.Any) -> None:
+        """
+        Setup the research tool VaRA with it's code base. This method sets up
```
Comment (boehmseb) [inline varats/tools/research_tools/phasar.py]:
```suggestion
        Setup the research tool phasar with it's code base. This method sets up
```

File: varats/tools/research_tools/research_tool.py [hunk 4]
```diff
@@ -297,8 +297,28 @@ def name(self) -> str:
     def is_build_type_supported(self, build_type: BuildType) -> bool:
         return build_type in self.__supported_build_types
 
+    @staticmethod
+    @abc.abstractstaticmethod
```
Comment (boehmseb) [inline varats/tools/research_tools/research_tool.py]: `@abc.abstractstaticmethod` is deprecated. Use `@staticmethod` with `@abstractmethod` instead.
Same goes for `@classmethod` and similar, if you should need them.

File: varats/utils/cli_util.py [hunk 5]
```diff
@@ -79,6 +82,29 @@ def initialize_logger_config() -> None:
     logging.basicConfig(level=log_level)
 
 
+ConcreteResearchTool = tp.TypeVar(
+    "ConcreteResearchTool", bound=ResearchTool[SpecificCodeBase]
+)
+
+
+def get_research_tool_type(name: str) -> tp.Type[ConcreteResearchTool]:
```
Comment (boehmseb) [inline varats/utils/cli_util.py]: According to https://github.com/python/mypy/issues/2756, this `TypeVar`s cannot depend on `TypeVar`s in the current type system of python. Replacing `SpecificCodeBase` with `CodeBase` or `Any` did not help either, as the types of the concrete research tools cannot be matched then.

The best solution I could find is:
```suggestion
def get_research_tool_type(name: str) -> tp.Type[ResearchTool[tp.Any]]:
```

File: varats/utils/cli_util.py [hunk 6]
```diff
@@ -95,15 +121,21 @@ def get_research_tool(
...
+    if not src_folder.exists():
+        src_folder.mkdir(parents=True)
+
+    return rs_type(src_folder)
 
 
 def get_supported_research_tool_names() -> tp.List[str]:
     """Returns a list of all supported research tools."""
-    return ["VaRA", "vara"]
+    return ["VaRA", "vara", "phasar"]
```
Comment (boehmseb) [inline varats/utils/cli_util.py]: I suggest to drop `VaRA` from this list. Instead, you could now specify a `key_normalizer` in the CLI-parser (s. https://github.com/madman-bob/python-argparse-utils). 

```suggestion
    return ["vara", "phasar"]
```

**Review by boehmseb — CHANGES_REQUESTED** [review]
Found a few places where you forgot to replace _vara_ and a few other small things.
Otherwise, this works fine for me.

**Review by boehmseb — APPROVED** [review]
lgtm

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- varats/tools/research_tools/phasar.py  6
         

Clones added
============
- varats/tools/research_tools/phasar.py  1
- varats/tools/research_tools/vara.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5612609&bid=18386596)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/tools/research_tools/phasar.py  6
         

Clones added
============
- varats/tools/research_tools/vara.py  1
- varats/tools/research_tools/phasar.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5612609&bid=18386596)

**Review by pdschubert — APPROVED** [review]
Looks good to me.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/tools/research_tools/phasar.py  6
         

Clones added
============
- varats/tools/research_tools/phasar.py  1
- varats/tools/research_tools/vara.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5612609&bid=18386596)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/tools/research_tools/phasar.py  6
         

Clones added
============
- varats/tools/research_tools/vara.py  1
- varats/tools/research_tools/phasar.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5612609&bid=18386596)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/tools/research_tools/phasar.py  6
         

Clones added
============
- varats/tools/research_tools/vara.py  1
- varats/tools/research_tools/phasar.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5612609&bid=18386596)

**Pull request merged**